### PR TITLE
Term threshold changes

### LINKED ISSUE #1059: Update Term Design and add Confidence Level
The current design [v0.0.20] related to terms has a couple of issues
1. Term durations are not strictly enforced, we can go back in time (upto a threshold level) to satisfy requirements related to a term duration. This means that the user is not clear what duration was considered to arrive at a recommendation
2. If enough data is not available that satisfies the entire term duration, then no recommendation is generated. eg, for short term, we need a full 96 data points with a measurement duration of 15 mins. This means that for any workload that is short lived, say 3 hours, we cannot generate any recommendations

To prevent this, we need to update the term design as follows
1. Fix term durations as follows (Assuming a measurement duration of 15 mins: short term = 24 hours (Max: 96 data points), medium term = 7 days (Max: 672  data points), long term = 15 days (Max: 1440 data points).
2. Remove buffer thresholds. This means that we do not go further back in time, beyond the prescribed term durations above, to get data for a given duration
3. Add a minimum data threshold. This represents the minimum amount of data needed for generating a recommendation associated with a given duration term. We can start with the following defaults. Minimum data points for: Short term: 30 mins (Min: 2 data points), Medium term: 2 days (Min: 192 data points), Long Term: 8 days (Min: 768 data points)
4. The design should allow for the minimum thresholds to be configurable for each individual experiment.
5. Add a confidence level associated with a recommendation. The confidence level is a value between 0 and 1 where 1 represents the highest confidence in the recommendation. The initial implementation will be based entirely on the number of data points used to generate the recommendation. It can be `confidence_level = actual_data_points / term_max_data_points`
6.

### PR DESCRIPTION
**Pull request by khansaad**
This PR partially fixes #1059  

- Added minimum data thresholds for each of the short, medium and long term.

### CODE REVIEW

**Comment by khansaad** [comment]
Test results attached

[PR-1081-remote_monitoring_tests.zip](https://github.com/kruize/autotune/files/13975596/PR-1081-remote_monitoring_tests.zip)
[remote_monitoring_tests.log](https://github.com/kruize/autotune/files/13975599/remote_monitoring_tests.log)

File: design/MonitoringModeAPI.md [hunk 1]
```diff
@@ -48,11 +48,13 @@ Below are the default terms used in resource analysis, along with their respecti
...
 
-| Term       | Duration | Threshold |
-|------------|----------|-----------|
-| Short      | 1 day    | 6 hours   |
-| Medium     | 7 days   | 6 hours   |
-| Long       | 15 days  | 6 hours   |
+| Term   | Minimum Data Threshold | Duration | Threshold |
+|--------|------------------------|----------|-----------|
+| Short  | 30 mins                | 1 day    | 6 hours   |
+| Medium | 2 Days                 | 7 days   | 6 hours   |
```
Comment (chandrams) [inline design/MonitoringModeAPI.md:54]: What does Threshold mean here?

File: tests/scripts/remote_monitoring_tests/helpers/utils.py [hunk 2]
```diff
@@ -302,10 +306,10 @@ def get_datetime():
 
 
 def term_based_start_time(input_date_str, term):
-    duration = {"short_term": 1, "medium_term": 7, "long_term": 15}
+    duration = {"short_term": 0.5, "medium_term": 48, "long_term": 192}
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/helpers/utils.py:309]: This should still remain 1 day, 7 days or 15 days for short term, medium and long term, as the min data points should be checked within this duration.

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py [hunk 3]
```diff
@@ -1032,43 +1030,45 @@ def test_list_recommendations_notification_codes(cluster_type: str):
             assert data['status'] == SUCCESS_STATUS
             assert data['message'] == UPDATE_RESULTS_SUCCESS_MSG
 
-            if j > 95:
-                response = update_recommendations(experiment_name, None, end_time)
-                data = response.json()
-                assert response.status_code == SUCCESS_STATUS_CODE
-                assert data[0]['experiment_name'] == experiment_name
+            response = update_recommendations(experiment_name, None, end_time)
+            data = response.json()
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:1034]: Also add an assert for the expected message to validate only the first data point doesn't have recommendations and rest has.

Comment (chandrams) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:1034]: I don't see all the test scenarios mentioned in the issue #1059 covered here, please add negative tests and tests for all the terms

**Comment by chandrams** [comment]
@khansaad - Recommendations aren't generated for long term when 768 data points or usage metrics are uploaded, can you please check.

File: tests/scripts/remote_monitoring_tests/helpers/utils.py (hunk 2 above)
Comment (chandrams) [reply tests/scripts/remote_monitoring_tests/helpers/utils.py:309]: Tests are failing as monitoring start time in the recommendations json is now changed. For short term, difference between end time & start time is 30 mins, for medium term 2 days and long term 8 days. Is this as expected?

```
    "monitoring_end_time": "2024-02-08T21:57:37.682Z",
      
                    "short_term": {
                      "duration_in_hours": 24.0,
                       "monitoring_start_time": "2024-02-08T21:27:37.682Z", 
                   
                    "medium_term": {
                      "duration_in_hours": 168.0,
                      "monitoring_start_time": "2024-02-06T21:57:37.682Z",
                      
                    "long_term": {
                      "duration_in_hours": 360.0,
                      "monitoring_start_time": "2024-01-31T21:57:37.682Z",
                      
```

**Comment by msvinaykumar** [comment]
> @khansaad - Recommendations aren't generated for long term when 768 data points or usage metrics are uploaded, can you please check.

@chandrams please retry with latest fix

**Review by msvinaykumar — APPROVED** [review]

File: tests/scripts/remote_monitoring_tests/helpers/utils.py (hunk 2 above)
Comment (chandrams) [reply tests/scripts/remote_monitoring_tests/helpers/utils.py:309]: @msvinaykumar - Is the monitoring_start_time as expected?

Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/helpers/utils.py:309]: No let me fix

**Comment by chandrams** [comment]
> > @khansaad - Recommendations aren't generated for long term when 768 data points or usage metrics are uploaded, can you please check.
> 
> @chandrams please retry with latest fix

Thanks @msvinaykumar, works fine now.

File: tests/scripts/remote_monitoring_tests/helpers/utils.py (hunk 2 above)
Comment (msvinaykumar) [reply tests/scripts/remote_monitoring_tests/helpers/utils.py:309]: Fixed now , Please chek

**Comment by chandrams** [comment]
@msvinaykumar @khansaad - monitoring_start_time fix didn't work for me, can you please check.

```
   "monitoring_end_time": "2024-01-29T15:29:19.770Z",
                  "current": {
                    "requests": {
                      "memory": {
                        "amount": 50.21,
                        "format": "MiB"
                      },
                      "cpu": {
                        "amount": 1.1,
                        "format": "cores"
                      }
                    },
                    "limits": {
                      "memory": {
                        "amount": 100.0,
                        "format": "MiB"
                      },
                      "cpu": {
                        "amount": 0.5,
                        "format": "cores"
                      }
                    }
                  },
                  "recommendation_terms": {
                    "short_term": {
                      "duration_in_hours": 0.5,
                      "notifications": {
                        "112101": {
                          "type": "info",
                          "message": "Cost Recommendations Available",
                          "code": 112101
                        },
                        "112102": {
                          "type": "info",
                          "message": "Performance Recommendations Available",
                          "code": 112102
                        }
                      },
                      "monitoring_start_time": "2024-01-29T14:59:19.770Z",
```

File: tests/scripts/remote_monitoring_tests/helpers/utils.py [hunk 4]
```diff
@@ -459,10 +462,18 @@ def validate_container(update_results_container, update_results_json, list_reco_
                             elif term == "long_term" and duration_in_hours > LONG_TERM_DURATION_IN_HRS_MAX:
                                 duration_in_hours = LONG_TERM_DURATION_IN_HRS_MAX
 
-                        print(
-                            f"Actual = {terms_obj[term]['duration_in_hours']} expected = {duration_in_hours}")
-                        assert terms_obj[term]["duration_in_hours"] == duration_in_hours, \
-                            f"Duration in hours did not match! Actual = {terms_obj[term]['duration_in_hours']} expected = {duration_in_hours}"
+                        print(f"Actual = {terms_obj[term]['duration_in_hours']} expected = {duration_in_hours}")
+
+                        if test_name is not None:
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/helpers/utils.py:467]: I think we can revert this change and test it.

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py [hunk 5]
```diff
@@ -20,6 +20,40 @@
     ("long_term_test_false", 15, long_term_list_reco_json_schema, LONG_TERM_DURATION_IN_HRS_MAX, False),
 ]
 
+term_input = [
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:23]: Include tests to validate short term recommendations not present and medium term recommendations are available and short term, medium term not present and long term is present.

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileInterface/ResourceOptimizationOpenshiftImpl.java [hunk 6]
```diff
@@ -145,7 +144,7 @@ public void generateRecommendation(KruizeObject kruizeObject, List<ExperimentRes
 
                         // Check for min data before setting notifications
                         // Check for at least short term
-                        if (!RecommendationUtils.checkIfMinDataAvailableForTerm(containerDataKruizeObject, RecommendationConstants.RecommendationTerms.SHORT_TERM)) {
+                        if (!RecommendationUtils.checkIfMinDataAvailableForTerm(containerDataKruizeObject, RecommendationConstants.RecommendationTerms.SHORT_TERM, monitoringEndTime)) {
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileInterface/ResourceOptimizationOpenshiftImpl.java:147]: It appears that this code is checking only for `SHORT_TERM` and if it is not available then setting the notification `INFO_NOT_ENOUGH_DATA`

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileInterface/ResourceOptimizationOpenshiftImpl.java]: Fixed now. Added test for the same as well.

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py (hunk 5 above)
Comment (khansaad) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:23]: Done

Comment (chandrams) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:23]: @khansaad - As discussed, please ensure all these combinations are covered, I will take a look too.

('L', 'M', 'S')
('L', 'M', 'NS')
('L', 'NM', 'S')
('L', 'NM', 'NS')
('NL', 'M', 'S')
('NL', 'M', 'NS')
('NL', 'NM', 'S')
('NL', 'NM', 'NS')

NL - No long term, NM - No medium term, NS - No short term

**Review by msvinaykumar — APPROVED** [review]

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py [hunk 7]
```diff
@@ -38,8 +38,26 @@
...
-    ("medium_term_test_192_data_points_non_contiguous", 192, medium_term_list_reco_json_schema, 48.0),
-    ("long_term_test_768_data_points_non_contiguous", 768, long_term_list_reco_json_schema, 192.0),
+    # contiguous scenarios
+    ("no_term_min_recomm", 1, None, 0, 0),
+    ("all_terms_min_recomm", 768, list_reco_json_schema, 192.0, 0),
+    ("only_short_term_min_recomm", 2, list_reco_json_schema, 0.5, 0),
+    ("only_medium_term_min_recomm", 192, medium_term_list_reco_json_schema, 48.0, 0),
+    ("only_long_term_min_recomm", 768, long_term_list_reco_json_schema, 192.0, 0),
+    ("short_and_medium_term_min_recomm", 192, list_reco_json_schema, 48.0, 0),
+    ("short_and_long_term_min_recomm", 768, list_reco_json_schema, 192.0, 0),
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:48]: Schema's need to be updated here. Existing schema's in the test may not work for all cases, we need to discuss on how to validate these and make changes later.

**Comment by chandrams** [comment]
@khansaad - Can you please share the test results.

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py (hunk 7 above)
Comment (khansaad) [reply tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py]: Updated and added new schema files

**Comment by khansaad** [comment]
> @khansaad - Can you please share the test results.
[remote_monitoring_tests.zip](https://github.com/kruize/autotune/files/14169156/remote_monitoring_tests.zip)

Added

**Review by chandrams — APPROVED** [review]
LGTM

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileInterface/ResourceOptimizationOpenshiftImpl.java [hunk 8]
```diff
@@ -144,16 +144,21 @@ public void generateRecommendation(KruizeObject kruizeObject, List<ExperimentRes
...
+                        if (!recommendationAvailable) {
                             RecommendationNotification recommendationNotification = new RecommendationNotification(
                                     RecommendationConstants.RecommendationNotification.INFO_NOT_ENOUGH_DATA
                             );
                             recommendationLevelNM.put(recommendationNotification.getCode(), recommendationNotification);
                             continue;
                         }
 
-                        boolean recommendationAvailable = false;
+                        recommendationAvailable = false;
```
[the remark this answers is not in the document]
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileInterface/ResourceOptimizationOpenshiftImpl.java:161]: This same variable `recommendationAvailable` is being used later on in the code to check for `cost` and `performance` recommendation engine.
Here: https://github.com/kruize/autotune/blob/da8e61d975c8adf899f3b7beb0c68df753433e6b/src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileInterface/ResourceOptimizationOpenshiftImpl.java#L318

File: tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py [hunk 9]
```diff
@@ -20,6 +26,60 @@
     ("long_term_test_false", 15, long_term_list_reco_json_schema, LONG_TERM_DURATION_IN_HRS_MAX, False),
 ]
 
+term_input = [
+    # test_name, num_res, reco_json_schema, expected_duration_in_hours, increment_end_time_by
+    ("short_term_test_2_data_points", 2, list_reco_json_schema, 0.5, 0),
+    ("medium_term_test_192_data_points", 192, medium_term_list_reco_json_schema, 48.0, 0),
+    ("long_term_test_768_data_points", 768, long_term_list_reco_json_schema, 192.0, 0),
+    ("long_term_test_769_data_points", 769, long_term_list_reco_json_schema, 192.3, 0),
+    ("long_term_test_800_data_points", 1440, long_term_list_reco_json_schema, 360.0, 0),
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/rest_apis/test_list_recommendations.py:35]: Replace 800 as 1440 in the test name

File: tests/scripts/remote_monitoring_tests/helpers/utils.py [hunk 10]
```diff
@@ -471,29 +474,14 @@ def validate_container(update_results_container, update_results_json, list_reco_
...
-                                    duration_in_hours = 192.0
-                            elif MEDIUM_AND_LONG in test_name:
-                                if term == MEDIUM_TERM:
-                                    duration_in_hours = 168.0
-                                else:
-                                    duration_in_hours = 192.0
-
-                            if MEDIUM_TERM in test_name and term == MEDIUM_TERM:
+
+                            if MEDIUM_TERM_TEST in test_name and term == MEDIUM_TERM:
```
Comment (chandrams) [inline tests/scripts/remote_monitoring_tests/helpers/utils.py:478]: Why are these conditions added back?

**Comment by dinogun** [comment]
@khansaad Please fix the conflicts

**Comment by khansaad** [comment]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Overall summary of the tests ~~~~~~~~~~~~~~~~~~~~~~~
Total time taken to perform the test 4214 seconds
Total Number of test suites performed 1
Total Number of tests performed 350
Total Number of tests passed 307
Total Number of tests failed 43

43 failed tests are expected due to create_exp validation issues. This will be taken care as part of #745 

Test results attached:
[remote_monitoring_tests.zip](https://github.com/kruize/autotune/files/14717499/remote_monitoring_tests.zip)

File: src/main/java/com/autotune/utils/KruizeConstants.java [hunk 11]
```diff
@@ -447,28 +450,46 @@ public static final class DurationAmount {
...
+                public static final double SHORT_TERM_TOTAL_DURATION_UPPER_BOUND_MINS = SHORT_TERM_MIN_DATA_THRESHOLD_MINS + BUFFER_VALUE_IN_MINS;
+                public static final double SHORT_TERM_TOTAL_DURATION_LOWER_BOUND_MINS = SHORT_TERM_MIN_DATA_THRESHOLD_MINS - BUFFER_VALUE_IN_MINS;
+
+                /* DATA POINTS */
+                public static final int SHORT_TERM_MAX_DATAPOINTS = 96;
+                public static final int SHORT_TERM_MIN_DATAPOINTS = 2;
+                public static final int MEDIUM_TERM_MAX_DATAPOINTS = 672;
+                public static final int MEDIUM_TERM_MIN_DATAPOINTS = 192;
+                public static final int LONG_TERM_MAX_DATAPOINTS = 1440;
+                public static final int LONG_TERM_MIN_DATAPOINTS = 768;
```
[the remark this answers is not in the document]
Comment (khansaad) [reply src/main/java/com/autotune/utils/KruizeConstants.java]: As discussed earlier in call, this was temporarily added and has been removed now. We aren't considering max and min datapoints.

File: src/main/java/com/autotune/utils/KruizeConstants.java [hunk 12]
```diff
@@ -458,23 +460,22 @@ private DurationAmount() {
...
+                        LONG_TERM_MIN_DATA_THRESHOLD_MINS - BUFFER_VALUE_IN_MINS;
+                /* MEDIUM TERM */
+                public static final double MEDIUM_TERM_TOTAL_DURATION_UPPER_BOUND_MINS = MEDIUM_TERM_MIN_DATA_THRESHOLD_MINS + BUFFER_VALUE_IN_MINS;
+                public static final double MEDIUM_TERM_TOTAL_DURATION_LOWER_BOUND_MINS = MEDIUM_TERM_MIN_DATA_THRESHOLD_MINS - BUFFER_VALUE_IN_MINS;
                 public static final double SHORT_TERM_HOURS = DurationAmount.SHORT_TERM_DURATION_DAYS * KruizeConstants.TimeConv.NO_OF_HOURS_PER_DAY;
                 public static final double MEDIUM_TERM_HOURS = DurationAmount.MEDIUM_TERM_DURATION_DAYS * KruizeConstants.TimeConv.NO_OF_HOURS_PER_DAY;
                 public static final double LONG_TERM_HOURS = DurationAmount.LONG_TERM_DURATION_DAYS * KruizeConstants.TimeConv.NO_OF_HOURS_PER_DAY;
+                public static final double SHORT_TERM_TOTAL_DURATION_UPPER_BOUND_MINS = SHORT_TERM_MIN_DATA_THRESHOLD_MINS + BUFFER_VALUE_IN_MINS;
+                public static final double SHORT_TERM_TOTAL_DURATION_LOWER_BOUND_MINS = SHORT_TERM_MIN_DATA_THRESHOLD_MINS - BUFFER_VALUE_IN_MINS;
```
Comment (dinogun) [inline src/main/java/com/autotune/utils/KruizeConstants.java:478]: Why do we still have buffer values? This is not needed anymore

**Review by dinogun — APPROVED** [review]
Approving based on test results
